crates/bevy_utils/src/anystack.rs

```diff
@@ -0,0 +1,218 @@
+struct AnyStackMeta {
+    offset: usize,
+    func: unsafe fn(value: *mut u8, user_data: *mut u8),
+}
+
+pub struct AnyStack {
+    bytes: Vec<u8>,
+    metas: Vec<AnyStackMeta>,
+}
+
+impl Default for AnyStack {
+    fn default() -> Self {
+        Self {
+            bytes: vec![],
+            metas: vec![],
+        }
+    }
+}
+
+unsafe impl Send for AnyStack {}
+unsafe impl Sync for AnyStack {}
+
+impl AnyStack {
+    #[inline]
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    #[inline]
+    pub fn len(&self) -> usize {
+        self.metas.len()
+    }
+
+    #[inline]
+    pub fn is_empty(&self) -> bool {
+        self.len() == 0
+    }
+
+    /// Clears in internal bytes and metas storage.
+    ///
+    /// # Safety
+    ///
+    /// This does not [`drop`] the pushed values.
+    /// The pushed values must be dropped via [`AnyStack::apply`] and
+    /// the provided `func` before calling this function.
+    #[inline]
+    pub unsafe fn clear(&mut self) {
+        self.bytes.clear();
+        self.metas.clear();
+    }
+
+    /// Push a new `value` onto the stack.
+    ///
+    /// # Safety
+    ///
+    /// `func` must safely handle the provided `value` bytes and `user_data` bytes.
+    /// [`AnyStack`] does not drop the contained member.
+    /// The user should manually call [`AnyStack::apply`] and in the implementation
+    /// of the provided `func` function from [`AnyStack::push`], the element should
+    /// be dropped.
+    pub unsafe fn push<U>(
+        &mut self,
+        value: U,
+        func: unsafe fn(value: *mut u8, user_data: *mut u8),
+    ) {
+        let align = std::mem::align_of::<U>();
+        let size = std::mem::size_of::<U>();
+
+        if self.is_empty() {
+            self.bytes.reserve(size);
+        }
+
+        let old_len = self.bytes.len();
+
+        let aligned_offset = loop {
+            let aligned_offset = self.bytes.as_ptr().add(old_len).align_offset(align);
+
+            if old_len + aligned_offset + size > self.bytes.capacity() {
+                self.bytes.reserve(aligned_offset + size);
+            } else {
+                break aligned_offset;
+            }
+        };
+
+        let offset = old_len + aligned_offset;
+        let total_bytes = size + aligned_offset;
+        self.bytes.set_len(old_len + total_bytes);
+
+        self.metas.push(AnyStackMeta { offset, func });
+
+        std::ptr::copy_nonoverlapping(
+            &value as *const U as *const u8,
+            self.bytes.as_mut_ptr().add(offset),
+            size,
+        );
+
+        std::mem::forget(value);
+    }
+
+    /// Call each user `func` for each inserted value with `user_data`.
+    ///
+    /// # Safety
```

Comment:
This operation should clear the stack.

If we become generic over the type of user_data, this operation can then also be safe.

(The user data should be passed as an `&mut T`)